src/pipewire.cpp

```diff
@@ -0,0 +1,187 @@
+    if (!b) {
+        std::cerr << "pipewire: out of buffers: " << strerror(errno) << std::endl;
+        return;
+    }
+
+    for (uint32_t i = 0; i < b->buffer->n_datas; ++i) {
+        struct spa_data *d = &b->buffer->datas[i];
+        memcpy(pr->buf + pr->buf_size, d->data, d->chunk->size);
+        pr->buf_size += d->chunk->size;
+        while (pr->buf_size >= pr->params.audio_frame_size) {
+            frame_writer->add_audio(pr->buf);
+            pr->buf_size -= pr->params.audio_frame_size;
+            if (pr->buf_size)
+                memmove(pr->buf, pr->buf + pr->params.audio_frame_size, pr->buf_size);
+        }
+    }
+
+    pw_stream_queue_buffer(pr->stream, b);
+}
+
+static const struct pw_stream_events stream_events = {
+    .version = PW_VERSION_STREAM_EVENTS,
+    .process = on_stream_process,
+};
+
+static void on_registry_global(void *data, uint32_t, uint32_t, const char *type, uint32_t, const struct spa_dict *props)
+{
+    PipeWireReader *pr = static_cast<PipeWireReader*>(data);
+
+    if (strcmp(type, PW_TYPE_INTERFACE_Node) != 0)
+        return;
+
+    const char *name = spa_dict_lookup(props, PW_KEY_NODE_NAME);
+    if (!name || strcmp(pr->params.audio_source, name) != 0)
+        return;
+
+    const char *media_class = spa_dict_lookup(props, PW_KEY_MEDIA_CLASS);
+    if (!media_class)
+        return;
+
+    pr->source_found = true;
+    pr->source_is_sink = strcmp(media_class, "Audio/Sink") == 0;
+}
+
+static const struct pw_registry_events registry_events = {
+    .version = PW_VERSION_REGISTRY_EVENTS,
+    .global = on_registry_global,
+};
+
+void PipeWireReader::start()
+{
+    struct pw_properties *props =
+        pw_properties_new(PW_KEY_MEDIA_TYPE, "Audio",
+                          PW_KEY_MEDIA_CATEGORY, "Capture",
+                          PW_KEY_MEDIA_ROLE, "Screen",
+                          PW_KEY_STREAM_CAPTURE_SINK, "true",
+                          PW_KEY_NODE_NAME, "wf-recorder",
+                          NULL);
+
+    if (params.audio_source) {
+        struct pw_registry *registry = pw_core_get_registry(core, PW_VERSION_REGISTRY, 0);
+        if (registry) {
+            struct spa_hook registry_listener;
+            pw_registry_add_listener(registry, &registry_listener, &registry_events, this);
+            seq = pw_core_sync(core, PW_ID_CORE, seq);
+            pw_thread_loop_wait(thread_loop);
+            if (!source_found) {
+                std::cerr << "pipewire: source " << params.audio_source << " not found, using default" << std::endl;
+            } else {
+                pw_properties_set(props, PW_KEY_STREAM_CAPTURE_SINK, source_is_sink ? "true" : "false");
+                pw_properties_set(props, PW_KEY_TARGET_OBJECT, params.audio_source);
+            }
+            spa_hook_remove(&registry_listener);
+            pw_proxy_destroy(reinterpret_cast<struct pw_proxy*>(registry));
+        }
+    }
+
+    stream = pw_stream_new(core, "wf-recorder3", props);
```

Comment:
Is "wf-recorder3" a typo or intentional for some reason?

Reply:
Copied it to match the pulse code, but not sure if it actually isn't a typo there.

https://github.com/ammen99/wf-recorder/blob/c43e8e11ec99e72d8c137fa2ea7e4ddb934275bd/src/pulse.cpp#L29-L30

Reply:
It probably could stand to be corrected in both places.

Reply:
I'd keep it to not break things.